### PR TITLE
random: mark RandAddPeriodic and SeedPeriodic as noexcept

### PR DESCRIPTION
**Pull request by fanquake**
The usage of `MilliSleep()` in SeedPeriodic (previously SeedSleep) was
[removed](https://github.com/bitcoin/bitcoin/pull/17270/commits/d61f2bb076d8f17840a8e79f1583d7f6e3e6d09a) in #17270, meaning it, and its users can now be marked `noexcept`.

This also corrects the docs in random.h for some of the changes in #17270.

### CODE REVIEW

File: src/random.cpp [hunk 1]
```diff
@@ -450,17 +450,7 @@ void LockingCallbackOpenSSL(int mode, int i, const char* file, int line) NO_THRE
...
- * cause a boost::thread_interrupted to be thrown.
- *
- * This means that SeedSleep, and all functions that invoke it are throwing.
- * However, we know that GetRandBytes() and GetStrongRandBytes() never trigger
- * this sleeping logic, so they are noexcept. The same is true for all the
- * GetRand*() functions that use GetRandBytes() indirectly.
- *
- * TODO: After moving away from interruptible boost-based thread management,
- * everything can become noexcept here.
+ * None of the RNG code should ever throw any exception.
```
Comment (laanwj) [inline src/random.cpp]: Very good :+1:

**Comment by practicalswift** [comment]
ACK 55b2cb199c276781b6daa5438af2da57dea3ac52

Very nice!

Somewhat related: it would be nice to have the same explicit guarantee for `LogPrint`/`LogPrintf`. That would allow compilers and static analyzers to treat a larger parts of our code base as "non-throwing" instead of "potentially throwing" :)

**Comment by laanwj** [comment]
> Somewhat related: it would be nice to have the same explicit guarantee for LogPrint/LogPrintf. 

Wouldn't that be easy to do? `LogPrintf` already catches the only internal exception, `tinyformat::format_error`, internally. I don't think any exceptions are supposed to leak out of it.

**Comment by practicalswift** [comment]
@laanwj Yes, exactly! It is only a matter of formalizing that with a `noexcept` :)

**Comment by MarcoFalke** [comment]
@practicalswift Wouldn't upstream be a better place to propose this change?

**Comment by practicalswift** [comment]
@MarcoFalke Upstream throws `tinyformat::format_error` which we catch: thus `noexcept` only makes sense on our end of things, no? :)

**Comment by MarcoFalke** [comment]
Fine, but `LogPrintStr` is also called, which might throw exceptions. In either case, this discussion should happen elsewhere.

**Comment by sipa** [comment]
ACK 55b2cb199c276781b6daa5438af2da57dea3ac52

**Review by sipa — COMMENTED** [review]
ACK 55b2cb199c276781b6daa5438af2da57dea3ac52

**Comment by laanwj** [comment]
I don't think `LogPrintStr` is supposed to ever raise an exception, but sure, that'd need to be checked. Agree this is off topic here.

**Comment by practicalswift** [comment]
@laanwj @MarcoFalke Please note that `SeedPeriodic` which is annotated `noexcept` in this PR calls `LogPrintf`:

https://github.com/bitcoin/bitcoin/blob/2065ef66ee6fb2b7bb442274f860813cad85b08c/src/random.cpp#L480-L491

Thus the question of a `noexcept` annotation for `LogPrintf` is not entirely off topic here, no? :)

**Comment by laanwj** [comment]
> Thus the question of a noexcept annotation for LogPrintf is not entirely off topic here :)

Sure. I'm going to work on a PR to make logging noexcept.

**Comment by sipa** [comment]
ACK. I accidentally redid part of this in #17670. The documentation improvements here should be done anyway, so let's do this PR first.

**Pull request merged**